### PR TITLE
initial test

### PR DESCRIPTION
**Pull request by JohnkDevHub**
AssetControllerGetAPITest

### CODE REVIEW

File: Service/pom.xml [hunk 1]
```diff
@@ -238,6 +238,11 @@
             <groupId>ch.qos.logback</groupId>
             <artifactId>logback-classic</artifactId>
         </dependency>
+        <dependency>
```
Comment (frankvanrietimqs) [inline Service/pom.xml:241]: Why on God's green earth would teh service be dependent on itself?

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import za.co.imqs.coreservice.dto.asset.AssetEnvelopeDto;
+import java.util.UUID;
+import static com.jayway.restassured.RestAssured.given;
+
+/**
+ * (c) 2020 IMQS Software
+ *<p>
+ * User: frankvr
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java:14]: You might want to fix that

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                then().assertThat().statusCode(HttpStatus.SC_OK);
+    }
+
+    @Test
+    public void assetGetNotFound()  throws Exception {
+        final UUID NONEXISTENT_ASSET = UUID.fromString("455ac960-8fc6-409f-b2ef-cd5be4ebe684");
+
+        given().
+                header("Cookie", login.getSession()).
+                get("/assets/{uuid}",NONEXISTENT_ASSET).
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java:59]: Note we add spaces after the comma

**Review by frankvanrietimqs — COMMENTED** [review]
Looks like you get the genarl idea. Maybe do a test where you create or update something next

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java (hunk 2 above)
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java:14]: okay

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java (hunk 3 above)
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetAPITest.java:59]: okay

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java [hunk 4]
```diff
@@ -17,9 +23,32 @@
...
+        prepPermissions();
+
+        deleteAssets(THE_ASSET);
+        createAsset();
+    }
+
     @Test
     public void addAssetDeleteLinkSuccess() {
-        fail("Not implemented");
+        given().
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java:37]: No that I see the names I gave these tests. They are pretty bad names.  `addAssetDeleteLinkSuccess` for instance should be called something like `addAssetLink` or `addLinkToAsse`t. There is no 'delete' here. Please update the names so they make sense.

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java [hunk 5]
```diff
@@ -42,6 +71,29 @@ public void addAssetDeleteForbidden() {
...
     public void addAssetDeleteNotFound() {
-        fail("Not implemented");
+
+        final String NONEXISTENT_EXTERNAL_ID = "c45036b1-a1fb-44f4-a254-a668c0d09zzz";
+        String nonexistent_id_type="c6a74a62-54f5-4f93-adf3-abebab3d3333";
+
+        given().
+                header("Cookie", login.getSession()).
+                delete("assets/link/{uuid}/to/{external_id_type}/{external_id}", THE_ASSET, nonexistent_id_type, NONEXISTENT_EXTERNAL_ID).
+                then().assertThat().statusCode(HttpStatus.SC_NOT_FOUND);
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java]: If you check the documentation you will see that 404 is not one of the possible return values. The server does not complain if you ask it to delete something that does not exist. 200 is the expected result.

**Review by frankvanrietimqs — CHANGES_REQUESTED** [review]

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java (hunk 4 above)
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java:37]: okay

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java (hunk 5 above)
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AssetControllerDeleteLinkAPITest.java]: okay

**Comment by JohnkDevHub** [comment]
I have read your observations from yesterday commits. changes will be made.

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByGroupingAPITest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                get("/assets/grouped_by/{grouping_id_type}/{grouping_id}", DTPW.GROUPING_TYPE_EMIS, THE_GROUPING_ID).
+                then().assertThat().statusCode(HttpStatus.SC_OK).body("size()",is(groupedAssetCount));
+    }
+
+    @Test
+    public void getAssetByGroupingNotFound()  {
+        // No Asset grouped
+
+        int groupedAssetCount=0;
+        given().
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByGroupingAPITest.java:73]: This is also a pretty meaningless test. Testing an empty database won't deliver false positives. I suggest  that you call getAssetByGroupingSuccess() to have a system populated with data and then try and get the non-existent asset to make sure

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByLinkAPITest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+        given().
+                header("Cookie", login.getSession()).
+                get("/assets/linked_to/{external_id_type}/{external_id}", external_id_type, THE_EXTERNAL_ID).
+                then().assertThat().statusCode(HttpStatus.SC_OK);
+    }
+
+    @Test
+    public void getAssetByLinkNotFound() {
+        // No Asset linked
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByLinkAPITest.java:59]: Again an empty dataset proves nothing

**Review by frankvanrietimqs — CHANGES_REQUESTED** [review]

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByGroupingAPITest.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        final String update_grouping_id = "25d0e46a-8360-4fc5-b792-994cd43311b5";
+
+        given().
+                header("Cookie", login.getSession()).
+                put("/assets/group/{uuid}/to/{grouping_id_type}/{grouping_id}", THE_OTHER_ASSET, DTPW.GROUPING_TYPE_EMIS, THE_GROUPING_ID).
+                then().assertThat().statusCode(HttpStatus.SC_CREATED);
+
+        assertLinked(THE_OTHER_ASSET,update_grouping_id);
+
+        int groupedAssetCount=2;
```
[the remark this answers is not in the document]
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByGroupingAPITest.java:61]: Okay

File: Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByGroupingAPITest.java (hunk 6 above)
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AssetControllerGetByGroupingAPITest.java:73]: okay

File: Service/src/test/java/za/co/imqs/api/asset/AbstractAssetControllerAPITest.java [hunk 9]
```diff
@@ -82,8 +82,16 @@ public Statement apply(Statement statement, org.junit.runner.Description descrip
     public static TestRule chain = RuleChain.outerRule(compose).around(login);
 
     public static final UUID THE_ASSET = UUID.fromString("455ac960-8fc6-409f-b2ef-cd5be4ebe683");
+    public static final UUID THE_SECOND_ASSET = UUID.fromString("455ac960-8fc6-409f-b2ef-cd5be4ebe684");
+    public static final UUID THE_THIRD_ASSET = UUID.fromString("455ac960-8fc6-409f-b2ef-cd5be4ebe685");
+    public static final UUID THE_FORTH_ASSET = UUID.fromString("455ac960-8fc6-409f-b2ef-cd5be4ebe686");
+    public static final UUID THE_FIRTH_ASSET = UUID.fromString("455ac960-8fc6-409f-b2ef-cd5be4ebe687");
```
Comment (frankvanrietimqs) [inline Service/src/test/java/za/co/imqs/api/asset/AbstractAssetControllerAPITest.java:88]: FIFTH

**Review by frankvanrietimqs — COMMENTED** [review]
Looks good, please proceed with the other tests

File: Service/src/test/java/za/co/imqs/api/asset/AbstractAssetControllerAPITest.java (hunk 9 above)
Comment (JohnkDevHub) [reply Service/src/test/java/za/co/imqs/api/asset/AbstractAssetControllerAPITest.java:88]: okay

**Review by frankvanrietimqs — APPROVED** [review]

**Pull request merged**